src/hotspot/share/opto/callGenerator.cpp

```diff
@@ -812,6 +816,646 @@ CallGenerator* CallGenerator::for_vector_reboxing_late_inline(ciMethod* method,
+               "unexpected address calculation shape");
+        array_cache_load_index_from_second_addp = array_cache_load_index_from_second_addp->in(1);
+        assert(!(array_cache_load_index_from_second_addp->Opcode() == Op_CastII &&
+                 array_cache_load_index_from_second_addp->in(0)->is_Proj() &&
+                 array_cache_load_index_from_second_addp->in(0)->in(0) == _cache_not_null_iff),
+               "no CastII because index_in_cache_array is known to be positive");
+#endif
+        index_in_cache_array = _kit.gvn().transform(new AddINode(array_cache_load_index_from_second_addp, index_in_cache_array));
+      }
+
+      if (_first_cache_probe_iff == nullptr) {
+        _first_cache_probe_iff = maybe_iff->as_If();
+        _first_index_in_cache = index_in_cache_array;
+      } else {
+        assert(_second_cache_probe_iff == nullptr, "no more than 2 cache probes");
+        _second_cache_probe_iff = maybe_iff->as_If();
+        _second_index_in_cache = index_in_cache_array;
+      }
+      return true;
+    }
+
+    // First traversal of the get() subgraph starts from the end of the method and follows control paths until it reaches
+    // the Thread.scopedValueCache() call. Given the shape of the method and some paths may have been trimmed and end with
+    // an uncommon trap, it could reach either the first or the second cache probe if first. Figure out which is the first
+    // here.
+    void adjust_order_or_first_and_second_probe_if(const Unique_Node_List &scoped_value_get_subgraph) {
+      if (_second_cache_probe_iff == nullptr) {
+        return;
+      }
+      assert(_first_cache_probe_iff != nullptr, "can't have a second iff if there's no first one");
+      Node_Stack stack(0);
+      stack.push(_cache_not_null_iff, 0);
+      while (stack.is_nonempty()) {
+        Node* c = stack.node();
+        assert(c->is_CFG(), "only cfg nodes");
+        uint i = stack.index();
+        if (i < c->outcnt()) {
+          stack.set_index(i + 1);
+          Node* u = c->raw_out(i);
+          if (scoped_value_get_subgraph.member(u) && u != c) {
+            if (u == _first_cache_probe_iff) {
+              return;
+            } else if (u == _second_cache_probe_iff) {
+              swap(_first_cache_probe_iff, _second_cache_probe_iff);
+              swap(_first_index_in_cache, _second_index_in_cache);
+              return;
+            }
+            stack.push(u, 0);
+          }
+        } else {
+          stack.pop();
+        }
+      }
+      fatal("should have found the cache probe ifs");
+    }
+
+    // ScopedValue.get() probes 2 cache locations. If, when pattern matching the get() subgraph, we found 2 ifs, then the
+    // first and second locations were probed. If the first if's other branch is to an uncommon trap, then that location
+    // never saw a cache hit. In that case, when the ScopedValueGetHitsInCacheNode is expanded, only code to probe
+    // the second location is added back to the IR.
```

Comment:
Can you say in the comment why that is ok, to remove the first probe location, and its trap?
I have a feeling it makes sense, but cannot say why...